lib/charms/observability_libs/v0/juju_topology.py

```diff
@@ -0,0 +1,230 @@
+        self.charm_name = charm_name
+        self.unit = unit
+
+    @classmethod
+    def from_charm(cls, charm):
+        """Creates a JujuTopology instance by using the model data available on a charm object.
+
+        Args:
+            charm: a `CharmBase` object for which the `JujuTopology` will be constructed
+        Returns:
+            a `JujuTopology` object.
+        """
+        return cls(
+            model=charm.model.name,
+            model_uuid=charm.model.uuid,
+            application=charm.model.app.name,
+            unit=charm.model.unit.name,
+            charm_name=charm.meta.name,
+        )
+
+    @classmethod
+    def from_dict(cls, data: dict):
+        """Factory method for creating `JujuTopology` children from a dictionary.
+
+        Args:
+            data: a dictionary with five keys providing topology information. The keys are
+                - "model"
+                - "model_uuid"
+                - "application"
+                - "unit"
+                - "charm_name"
+                `unit` and `charm_name` may be empty, but will result in more limited
+                labels. However, this allows us to support payload-only charms.
+
+        Returns:
+            a `JujuTopology` object.
+        """
+        return cls(
+            model=data["model"],
+            model_uuid=data["model_uuid"],
+            application=data["application"],
+            unit=data.get("unit", ""),
+            charm_name=data.get("charm_name", ""),
+        )
+
+    def as_dict(
+        self,
+        *,
+        remapped_keys: Dict[str, str] = {},
+        excluded_keys: List[str] = [],
+        uuid_length: Optional[int] = None,
+    ) -> OrderedDict:
+        """Format the topology information into an ordered dict.
+
+        Keeping the dictionary ordered is important to be able to
+        compare dicts without having to resort to deep comparisons.
+
+        Args:
+            remapped_keys: A dictionary mapping old key names to new key names,
+                which will be substituted when invoked.
+            excluded_keys: A list of key names to exclude from the returned dict.
+            uuid_length: The length to crop the UUID to.
+        """
+        model_uuid = self.model_uuid[: uuid_length - 1] if uuid_length else self.model_uuid
+        ret = OrderedDict(
+            [
+                ("model", self.model),
+                ("model_uuid", model_uuid),
+                ("application", self.application),
+                ("unit", self.unit),
+                ("charm_name", self.charm_name),
+            ]
+        )
+
+        for exclusion in excluded_keys:
+            if exclusion in ret:
+                ret.pop(exclusion)
+
+        if remapped_keys:
+            ret = OrderedDict(
+                (remapped_keys.get(k), v) if remapped_keys.get(k) else (k, v) for k, v in ret.items()  # type: ignore
+            )
+
+        return ret
+
+    def as_identifier(self) -> str:
```

Comment:
`excluded_keys=["unit", "charm_name"]` and `uuid_length=8` is very opinionated.
Is there a better name for this than `identifier`, e.g. `app identifier`? Or perhaps just let users extend? Or perhaps:

```suggestion
    def app_identifier(self, excluded_keys=("unit", "charm_name"), uuid_length=8) -> str:
```

Reply:
Yeah, I'm totally fine with making them properties instead. I wanted to make the API consistent, and opted for dropping properties altogether. I can just as well see us going the other direction and making all methods except the class methods `@properties`, but then I think it would make sense to also drop the `as_` prefix.

Reply:
This lib is called observability_libs so we can totally do our opinionated contracts here.
However, it may hamper the adoptability of this lib outside of o11y charms.

Reply:
Sure, it's observability libs, but we also want to try to get this into "core" OF, and a good contract will help there.

Beyond that, we want to make sure that the things we are actually observing set their labels in a way that they're observable, right? A very strict contract is ideal there, and this contract is not strictly tied to observability except in the sense that users may want to observe their charms, and these are useful labels. It could be equally useful for geo-replicating Ceph/Gluster, determining locality for distributed databases, routing endpoints to closest-hop like BGP, etc. 

Imagine an "integrator charm" into which an admin can pass a dict/map which correlates controllers/clouds/AZs/datacenters/whatever and models so a highly distributed database/object store/whatever can effectively route clients to the closest endpoint. This class can provide that, and it is not specific to observability. Will anyone actually do this? No idea. But keeping a very strict definition of what these values mean will certainly help, since users won't have have to try to guess what values are in "topology" from a charm they didn't write.

I suggested `@property` mostly because then you don't need `()` to call a method which takes no arguments of any kind, but if `model_uuid`, for example, was also:

```python
@property
def model_uuid(self) -> str:
    return self._model_uuid
```

It would also make it more challenging to accidentally (or intentionally) reassign it. Sure, you could just  `topology._model_uuid = ...`, but "Pythonic" practice is basically "any `attr` (which includes methods) which is not prepended by `_` is public". 

I mean, it's actually public anyway, but think of it like exporting a field from a struct in Go. If you intend for it to be "internal", add `_`, or it will definitely pop up in autocompletion for anything which uses LSP (pycharm, emacs, jedi, neovim, vscode, etc), which makes the surface area we want to expose muddier.

Reply:
Implemented the getters.